gax-java/gax/src/test/java/com/google/api/gax/tracing/MetricsTracerTest.java

```diff
@@ -0,0 +1,263 @@
+
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "INVALID_ARGUMENT",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(1, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testAttemptCancelled_recordsAttributes() {
+    // initialize mock-request
+    Object mockCancelledRequest = new Object();
+    // Attempt #1
+    metricsTracer.attemptStarted(mockCancelledRequest, 0);
+    metricsTracer.attemptCancelled();
+
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "CANCELLED",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(1, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testAttemptFailedRetriesExhausted_recordsAttributes() {
+    // initialize mock-request
+    Object mockRequest = new Object();
+    // Attempt #1
+    metricsTracer.attemptStarted(mockRequest, 0);
+    ApiException error0 =
+        new DeadlineExceededException(
+            "deadline exceeded", null, new FakeStatusCode(Code.DEADLINE_EXCEEDED), false);
+    metricsTracer.attemptFailedRetriesExhausted(error0);
+
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "DEADLINE_EXCEEDED",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(1, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testAttemptPermanentFailure_recordsAttributes() {
+
+    // initialize mock-request
+    Object mockRequest = new Object();
+    // Attempt #1
+    metricsTracer.attemptStarted(mockRequest, 0);
+    ApiException error0 =
+        new NotFoundException("not found", null, new FakeStatusCode(Code.NOT_FOUND), false);
+    metricsTracer.attemptFailedRetriesExhausted(error0);
+
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "NOT_FOUND",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(1, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testAddAttributes_recordsAttributes() {
+
+    metricsTracer.addAttributes("FakeTableId", "12345");
+    Truth.assertThat(metricsTracer.getAttributes().get("FakeTableId").equals("12345"));
+  }
+
+  @Test
+  public void testExtractStatus_errorConversion_apiExceptions() {
+
+    // test all cases of ApiExceptions
+    for (Code code : Code.values()) {
```

Comment:
We don't have to test all codes, using one code is good enough. Because we don't differentiate the statuses in this method.

Reply:
understood, changed to just one code.